### DISCUSSION TITLE
[Feature Request] Add gamepad support to UI

### DISCUSSION THREAD

**Discussion by Buri** [lutris/lutris#4012] [Ideas]
Currently, Lutris has no support for gamepad input. This makes it very difficult to use on handheld devices like Aya Neo or Steam Deck.

This would probably mean adding support for controller intself, adding appropriate controller inputs (in case of support for different controller models), create system for navigation in UI and support for on-screen keyboard.

**Comment by DavidoTek** [comment]
Could probably be done with the [inputs](https://pypi.org/project/inputs) library.
Did something similar with [ProtonUp-Qt](https://github.com/DavidoTek/ProtonUp-Qt) where I emulated keyboard input, though Lutris is using GTK instead of Qt. Only problem with inputs is that it is quite outdated and has a few bugs...

+1 Would be quite useful for the Steam Deck too.

EDIT: Navigation could be similar to that of [HGL 2.1.0](https://github.com/Heroic-Games-Launcher/HeroicGamesLauncher/releases/tag/v2.1.0)

**Comment by strycore** [comment]
The plan is to have an UI made with Godot that can be used with a gamepad. 
Having gamepad support on the main UI would be useful too, since we're not going to recreate the whole lutris UI in Godot (at least not for the foreseeable future)

At some point, I'd also like to pick up on the development of SC Controller and try to integrate it with Lutris. That would allow to have controller support for all games, not only Steam ones.

**Reply by Buri** [reply]
So how would this work? Make standalone app in godot that would use Lutris libs internally/launch Lutris app in background? Or is there a way to integrate Godot into existing python project?

The former should be pretty "easy" to do, not sure if the latter is even possible.

**Reply by Buri** [reply]
I threw together quick mockup how it might work: https://github.com/Buri/lutris-gamepad-ui
Good news is that default focus behaviour seems to work well ootb, so I will continue investigating further.

**Reply by DavidoTek** [reply]
> is there a way to integrate Godot into existing python project

Godot has Python bindings: https://github.com/touilleMan/godot-python
It will include a Python interpreter and all libraries withing the Godot project. That would probably mean that the Godot UI would be completly separate from "normal" Lutris and just replacing GTK with Godot.  
Ideally you will be able to launch normal Lutris/advanced dialogs from the Godot UI, not sure if this would be possible then.

EDIT: Made a mockup image, should be quite easily possible with Godot.
![lutrisgodotui-mockup1](https://user-images.githubusercontent.com/54072917/151262317-9cc97e33-e329-4335-acea-51ac388cfc1d.png)

**Reply by strycore** [reply]
I'm not sure Python bindings would benefit us at all here. It's unlikely that we would be able to share the same code base between a Godot and a Gtk project. I won't make the same mistake again of choosing some python bindings for a library only to be treated as a second class citizen as a result. We only need the Godot UI and the Lutris client to be able to communicate in some way, this doesn't require to share a common programming language.

Note that all config dialogs and installer dialogs are not Gtk.Dialog instances but Gtk.Window instances. This means we are not dependent on the main window to display them. We already can launch installs via a command line, it wouldn't be too hard to launch a config dialog for a game.

**Reply by Buri** [reply]
> EDIT: Made a mockup image, should be quite easily possible with Godot. ![lutrisgodotui-mockup1](https://user-images.githubusercontent.com/54072917/151262317-9cc97e33-e329-4335-acea-51ac388cfc1d.png)

This looks good, but I will make few changes: 
* Move the left menu to the top. This way it will be easy to map to L1/R1 buttons and not confuse the users
* Move the bottom buttons into a menu, so using directional keys stays only for games unless user explicitly changes his workflow

Please note I am not an UX designer, but I do some experience developing ui's, so everithing in regard is very much open for discussion.

EDIT: I will also try getting the demo built for web so it can be easily evaluated.

**Reply by strycore** [reply]
I've run @Buri's mockup in Godot and with @DavidoTek 's mockup, we're all agreeing on the same type of design.

I want to be able to ship this as quickly as possible so we should really restrain the features to the absolute minimum. Filter and search are ok since that's just several ways of querying the database. Beside that, I don't think there will be a lot more features on the initial version.

That said, we have total freedom on how to use this information and display it in ways that aren't possible with Gtk, at least not easily.

**Reply by Buri** [reply]
>We only need the Godot UI and the Lutris client to be able to communicate in some way

Do you have any particular way in mind?
> We already can launch installs via a command line, it wouldn't be too hard to launch a config dialog for a game.

This would require additional changes on the main client, right?
EDIT: formatting

**Reply by strycore** [reply]
@Buri something I've noted in your mockup is that it tries to replicate some aspects of our current UI, like loading the full game list and displaying it in a scrolling window. One of the benefits of using Godot is that we can move away from this kind of paradigm and build a much more responsive UI. 

What I'd like to achieve is something that loads a small number of games at a time, and when the user scrolls through the list, it gets dynamically loaded. It's not easily doable in GTK, but in Godot I can see it happening quite well.

**Reply by strycore** [reply]
So far I haven't looked into the options for communicating between Godot and Lutris but we have a bunch of options... DBus, sockets, a HTTP service... 

We'll have to do some heavy lifting on the client part to adapt for another UI, that will likely be the point where we release Lutris 0.6

**Reply by Buri** [reply]
> something I've noted in your mockup is that it tries to replicate some aspects of our current UI, like loading the full game list and displaying it in a scrolling window. One of the benefits of using Godot is that we can move away from this kind of paradigm and build a much more responsive UI

My idea was to load X games needed for first screen/two screens and load rest on demand, but still show the scrollbar so user sees that there is more. We could also go for paging or any other solution, Im open to ideas.

**Reply by strycore** [reply]
We can load the total number of games so we can still show a scrollbar that looks correct. The important thing is to not load all the images all at once like we do now. We can preload 1 or 2 pages forward and backward and load the rest when triggered by the user.

**Reply by Buri** [reply]
I looked into network options in Godot - it has native support support for http(s) and websockets. I also tried to create custom bindings for dbus and got demo working without too much trouble. There is pretty easy bindings library for Rust, so we can create native modules easily.

@DavidoTek is also pushing updates on the gui side, so we are progressing nicely with the prototype.

**Reply by Buri** [reply]
We have new update today. More styling work and I also added sample image loading for game artwork. This unfortunately doesnt work on web for some reason, but works no problem on desktop.

**Comment by Buri** [comment]
I've added automatic builds, so latest version is always available on: https://buri.github.io/lutris-gamepad-ui/

**Reply by awsms** [reply]
Any installation/usage instructions? :)

**Reply by DavidoTek** [reply]
You can either use the web version or clone the repo from [here](https://github.com/Buri/lutris-gamepad-ui), but there isn't any real functionality as of now, it's more a mockup.

**Comment by DavidoTek** [comment]
I had a look at the DBus specification. We should agree on a standard way for communicating with the Lutris launcher.

Maybe like this? https://gist.github.com/DavidoTek/41945c60dfe82e93efb4eca0758a1e4b

For HTTP: https://gist.github.com/DavidoTek/76ca38f406edcf1f73c362c0141fa9e9

**Reply by Buri** [reply]
I think we should first decide wheter we even want to use DBus in the first place. I personally have no experience with it. I have also no idea how much work it would be to integrate to Lutris.
So far we have these options:
* HTTP client/server (native support in godot)
* DBus
* local socket communication
I was also looking on gRPC, but again, I have no personal experience. This might be good place for maintainers to chip in if they have any preference. @strycore ?

**Reply by DavidoTek** [reply]
Another option would be to use WebSockets and JSON. Both are supported by Python and Godot and quite simple to work with.
Could look something like this: https://gist.github.com/DavidoTek/ce875d88f06833e83515838c8051cdbb

**Reply by ffiarpg** [reply]
Might be a dumb idea but any chance of integrating with this project for UI?
https://github.com/retrohub-org/retrohub

**Reply by strycore** [reply]
Not a dumb idea. I didn't know about Retrohub. I was planning to have better integration with OpenGamepadUI which is based on Godot 4 as well.

**Reply by ffiarpg** [reply]
Or this project in Godot 4 https://github.com/ShadowBlip/OpenGamepadUI
Which is already developing a lutris plugin.
https://github.com/ShadowBlip/OpenGamepadUI-lutris

**Comment by andrew-ld** [comment]
Hi,
I’ve created a gamepad-friendly UI that allows you to launch Lutris games.

https://github.com/andrew-ld/lutris-gamepad-ui

**Comment by Jopp-gh** [comment]
*Old topic, same need.*

### Gamepad benefit
Gamepads are very comfortable in contrast to fast-finger acrobatics on keyboard (eg aRPG titles in ego view, 3rd person-view). 

### Keyboard input
Also, pressing and keep pressed buttons like wasd is tedious, a thumbstick instead, is just a one-lever button with multi directional output, much faster, less hectic than reaching multiple buttons in diff positions.

### Remapping software 
Remap software can be buggy too but apart this,   gamepads are hard to configure because:
1. games recognize plugged gamepads so games force you to use their fixed gamepad presets 
2. custom remaps will mix with your default gamepad presets leading to unpredictable results

### Problems with Remapping software
Always talking of gamepads: if you're lucky, your custom keys "play" in game otherwise you will get a mix where default gamepad presets specified by your game will override your custom keys or worse, play  both keys one after another. 

### Feature request
My wish would be to find a configurable remap menu for keyboards and gamepads build-in into Lutris with following features:
- a key remap layer able to deactivate/override games' gamepad and keyboard presets
- add a quick-configuration wizard (you press a key on the gamepad, then the equivalent on keyboard, etc, till all keys are set)
- a way to add multiple events to keys, causing to release diff actions: long press and release, long press and keep, double press, press the same key repeatedly to switch-loop through diff keys.

The benefit of introducing  gamepad support would be to have all important features in one, useful meaningful great launcher and manager. No more additional software to run in tandem with Lutris.

#### Foot note
I agree, configuring gamepads isn't always that simple because people love easy, quick presets and tinkering how to best configure a new remap can be quite a study. Nevertheless, gamepads are  beneficial for hand health and aren't just some luxury/freak toys.

**Comment by strycore** [comment]
> Old topic, same need.

This is a very different need, but luckily, it's also on the roadmap. What you want is an integration with [InputPlumber](https://github.com/ShadowBlip/InputPlumber/) and a GUI to configure it. I've already done this once [for another project](https://github.com/playtron-os/gamelab) 

> * games recognize plugged gamepads so games force you to use their fixed gamepad presets
> * custom remaps will mix with your default gamepad presets leading to unpredictable results
> * Always talking of gamepads: if you're lucky, your custom keys "play" in game otherwise you will get a mix where default gamepad presets specified by your game will override your custom keys or worse, play both keys one after another.

Not if you're using InputPlumber.

> No more additional software to run in tandem with Lutris.

Well you'd still need InputPlumber, but that runs in the background so you would not interact with it directly

**Reply by Jopp-gh** [reply]
Thanks for your answer, @strycore !

### InputPlumber
**InputPlumber** seems to be a newer development? 

Anyway, if InputPlumber is a headless process, then some GUI app could add **multiple key events**:
Since gamepads are very limited in number of buttons (which is also their benefit) expanding with multiple events would be "pretty much a must". 


### Multiple key events
This feature is indeed part of *advanced configuration*, but feels rather like a natural extension than some new, additional feature.

### Workaround
Right now I'm using **AntimicroX**, this is  great software; maybe its GUI is a bit confusing but it delivers. 

Lutris supports automatic loading of AntimicroX's remap profiles on game launch, this fits perfectly and is a great win! I guess, also InputPlumber is unable to choose the right remap profile by itself, need to check. 

### Additional notes
With AntimicroX, keyboard configuration isn't an issue but with gamepads, AntimicroX gives me discrete outputs, some are mixed.

**Reply by strycore** [reply]
are you a bot?

**Reply by Jopp-gh** [reply]
Seems like this? Just trying to highlight the importance of dynamic key events. 

Next time I won't add headers. The text is a bit long but the intent was to be as clear as possible.
